File: app/javascript/channels/chatroom_channel.js
import consumer from "channels/consumer"

consumer.subscriptions.create("ChatroomChannel", {
  connected() {
    console.log("connected to chatroom channel");
  },

  disconnected() {
    console.log("disconnected from chatroom channel");
  },

  received(data) {
    const messageContainer = document.getElementById("message-container");

    if (data.type === "connected" || data.type === "disconnected") {
      updateOnlineMembersList( data.data.online_users );
    } else {
      // Handle new messages
      messageContainer.insertAdjacentHTML('beforeend', data.mod_message);
      messageContainer.scrollTop = messageContainer.scrollHeight;

      // Clear the text field
      const chatMessageBody = document.getElementById("chat-message-body");
      chatMessageBody.value = "";
    }
  }
});

function updateOnlineMembersList(data) {
  // This function will dynamically update the online members list
  const onlineMembersContainer = document.getElementById("online-members");
  onlineMembersContainer.innerHTML = ""; // Clear the current list

  data.forEach(user => {
    const userElement = document.createElement("p");
    userElement.className = "text-white p-2 font-medium text-lg border border-b-gray-500";
    userElement.textContent = user;
    onlineMembersContainer.appendChild(userElement);
  });
}

File: app/channels/chatroom_channel.rb
# app/channels/chatroom_channel.rb
class ChatroomChannel < ApplicationCable::Channel
  @@online_users = []

  def subscribed
    # Track the current user
    unless @@online_users.include?(current_user.username)
      @@online_users << current_user.username
      broadcast_user_status("connected")
    end

    stream_from "chatroom_channel"
  end

  def unsubscribed
    # Remove the user from the online users list when they leave
    if @@online_users.delete(current_user.username)
      broadcast_user_status("disconnected")
    end
  end

  private

  def broadcast_user_status(status)
    ActionCable.server.broadcast "chatroom_channel", {
      type: status,
      data: { online_users: @@online_users }
    }
  end
end
